repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go

```diff
@@ -0,0 +1,312 @@
+}
+
+func initTempNode(ctx context.Context, bootstrap []string, peers []peer.AddrInfo) (string, error) {
+	err := setupPlugins()
+	if err != nil {
+		return "", err
+	}
+
+	identity, err := config.CreateIdentity(ioutil.Discard, []options.KeyGenerateOption{
+		options.Key.Type(options.Ed25519Key),
+	})
+	if err != nil {
+		return "", err
+	}
+	cfg, err := config.InitWithIdentity(identity)
+	if err != nil {
+		return "", err
+	}
+
+	// create temporary ipfs directory
+	dir, err := ioutil.TempDir("", "ipfs-temp")
+	if err != nil {
+		return "", fmt.Errorf("failed to get temp dir: %s", err)
+	}
+
+	// configure the temporary node
+	cfg.Routing.Type = "dhtclient"
+
+	// Disable listening for inbound connections
+	cfg.Addresses.Gateway = []string{}
+	cfg.Addresses.API = []string{}
+	cfg.Addresses.Swarm = []string{tempNodeTcpAddr}
+
+	if len(bootstrap) != 0 {
+		cfg.Bootstrap = bootstrap
+	}
+
+	if len(peers) != 0 {
+		cfg.Peering.Peers = peers
+	}
+
+	err = fsrepo.Init(dir, cfg)
+	if err != nil {
+		os.RemoveAll(dir)
+		return "", fmt.Errorf("failed to initialize ephemeral node: %s", err)
+	}
+
+	return dir, nil
+}
+
+func (f *IpfsFetcher) startTempNode(ctx context.Context) error {
+	// Open the repo
+	r, err := fsrepo.Open(f.ipfsTmpDir)
+	if err != nil {
+		return err
+	}
+
+	// Create a new lifetime context that is used to stop the temp ipfs node
+	ctxIpfsLife, cancel := context.WithCancel(context.Background())
+
+	// Construct the node
+	node, err := core.NewNode(ctxIpfsLife, &core.BuildCfg{
+		Online:  true,
+		Routing: libp2p.DHTClientOption,
+		Repo:    r,
+	})
+	if err != nil {
+		cancel()
+		r.Close()
+		return err
+	}
+
+	ipfs, err := coreapi.NewCoreAPI(node)
+	if err != nil {
+		cancel()
+		return err
+	}
+
+	stopFunc := func() {
+		// Tell ipfs to stop
+		cancel()
+		// Wait until ipfs is stopped
+		<-node.Context().Done()
+
+		fmt.Println("migration peer", node.Identity, "shutdown")
+	}
+
+	addrs, err := ipfs.Swarm().LocalAddrs(ctx)
+	if err != nil {
+		// Failure to get the local swarm address only means that the
+		// downloaded migrations cannot be fetched through the temporary node.
+		// So, print the error message and keep going.
+		fmt.Fprintln(os.Stderr, "cannot get local swarm address:", err)
+	}
+
+	f.addrInfo = peer.AddrInfo{
+		ID:    node.Identity,
+		Addrs: addrs,
+	}
+
+	f.ipfs = ipfs
+	f.ipfsStopFunc = stopFunc
+
+	return nil
+}
+
+func parsePath(fetchPath string) (ipath.Path, error) {
+	ipfsPath := ipath.New(fetchPath)
+	if ipfsPath.IsValid() == nil {
+		return ipfsPath, nil
+	}
+
+	u, err := url.Parse(fetchPath)
+	if err != nil {
+		return nil, fmt.Errorf("%q could not be parsed: %s", fetchPath, err)
+	}
+
+	switch proto := u.Scheme; proto {
+	case "ipfs", "ipld", "ipns":
+		ipfsPath = ipath.New(path.Join("/", proto, u.Host, u.Path))
+	default:
+		return nil, fmt.Errorf("%q is not an IPFS path", fetchPath)
+	}
+	return ipfsPath, ipfsPath.IsValid()
+}
+
+func setupPlugins() error {
```

Comment:
It does work.  Since the temporary download node is using the same code as the daemon, any plugins that work for one will work for the other. We do not load the plugins twice because the error that results in handled.

Is it necessary?... Probably only for testing.  As long as the IpfsFetcher is used from within a running ipfs daemon, this is not needed as the plugins are already loaded.  If we wanted to use IpfsFetcher outside of the daemon, then we do need to load the plugins, or we get an error like `failed to initialize ephemeral node: unknown datastore type: flatfs`.

Calling `setupPlugins` when plugins are already loaded seems harmless, but should not be necessary if it is only ever called from within the daemon.  Therefore, I moved it out to the test code.